File: c++/source/parser.cpp
#include "parser.hpp"
#include "tagset.hpp"

//Error messages
const char* SECTION_EOF="Reached end of string/file while parsing a section.";
const char* UNQUOTED_EOF=
	"Reached end of string/file while parsing unquoted text.";
const char* QUOTED_EOF="Reached end of string/file while parsing quoted text.";
const char* EXPECTED_CLOSE_SECTION="Expected close bracket ].";
const char* TAG_EOF="Reached end of string/file while parsing a tag";
const char* UNEXPECTED_CLOSE_SECTION=
	"Unexpected close bracket ]. Did you forget to close a tag?";
const char* UNNAMED_ATTR="Forgot to assign a value to the named attribute.";
const char* ILLEGAL_NAMED="The colon is disallowed in unquoted text.";
const char* UNEXPECTED_SPACE="Expected a value, found whitespace. "
	"There's a problem with the API's parser code.";
const char* COSMIC_RAY_ERR="Something went horribly wrong. "
	"There's a problem with the API's parser code.";

const char* UNICODE_EOF="Reached end of string/file while parsing a Unicode "
	"codepoint.";

//Relevant characters
char OPEN_TAG='{';
char NAMED_ATTR=':';
char CLOSE_TAG='}';
char OPEN_SECTION='[';
char CLOSE_SECTION=']';
char QUOTE1='"';
char QUOTE2='\'';
char QUOTE3='`'
char ESCAPE_CHAR='\\';

/**
 * Safe pointer class based on RAII. This simplifies the logic and increases
 *  the safety of deallocating flakes when an ParseError occurs.
**/
template<typename T>
class SafePointer{
	protected:
		T* fp;
	public:
		SafePointer(T* p=NULL):fp(p){}

		~SafePointer(){
			if(fp){
				delete fp;
			}
		}

		T* operator ->(){
			return fp;
		}

		operator bool(){
			return (bool)fp;
		}

		/**
		 * Yield ownership of the pointer to the caller.
		**/
		T* yield(){
			T* p=fp;
			fp=NULL;
			return p;
		}

		/**
		 * Claim ownership over the given pointer.
		**/
		T* claim(T* p){
			if(fp){
				delete fp;
			}
			return fp=p;
		}

		/**
		 * Return the pointer without changing ownership (used for passing to
		 *  functions which don't do anything to it).
		**/
		T* value(){
			return fp;
		}
};

ParseError::ParseError(const std::string& s,size_t line,size_t col):
	std::runtime_error(s),line(line),col(col){}

bool Parser::maybe_char(char c){
	if(it!=end and *it==c){
		++it;
		//Maybe is never used with whitespace, so it doesn't have to worry
		//about calculating line changes.
		++col;
		return true;
	}
	return false;
}

void Parser::calc_line(char c){
	if(c=='\r'){
		if(it!=end and *it=='\n'){
			++it;
		}
		++line;
		col=0;
	}
	//Unix-style and some ASCII newlines in Unicode standard
	else if(c=='\n' or c=='\v' or c=='\f'){
		++line;
		col=0;
	}
	else{
		++col;
	}
}

char Parser::parse_escape(char c,const char* eof){
	if(c==ESCAPE_CHAR){
		//Escapes consume the next character regardless of what it is
		if(++it!=end){
			//We still need to check for newlines for error messages.
			calc_line(*it);
			if(++it!=end){
				return *it;
			}
		}
		
		if(eof){
			throw ParseError(eof,line,col);
		}
		
		return *(--it);
	}
	
	calc_line(c);
	
	//Can't return c directly because calc_line may have incremented.
	return *it;
}

void Parser::space(){
	char c;
	while(it!=end and isspace(c=*it)){
		++it;
		calc_line(c);
	}
}

Tag* Parser::parse_tag(){
	if(!maybe_char(OPEN_TAG)){
		return NULL;
	}
	space();

	SafePointer<Tag> tag(new Tag());

	while(!maybe_char(CLOSE_TAG)){
		SafePointer<Flake> key(parse_value());
		space();
		if(maybe_char(NAMED_ATTR)){
			colonline=line;
			coloncol=col-1;

			space();
			SafePointer<Flake> val(parse_value());
			space();

			//This isn't owned, so no safe pointer
			if(Flake* f=tag->get(key.value())){
				if(Section* s=f->as_section()){
					//Merge the value with the section.
					s->add(val.yield());
				}
				else{
					//Merge the values into a section.
					tag->set(key.value(),new Section(f,val.yield()));
				}
				//key has no purpose, so it'll be deleted
			}
			else{
				//Add a new named attribute.
				tag->set(key.yield(),val.yield());
			}
		}
		else{
			//Add the value to the positional attributes.
			tag->add(key.yield());
		}
	}

	if(ts){
		if(Tagdef td=ts->get(tag->get((size_t)0))){
			return td(tag.yield())->as_tag();
		}
	}
	return tag.yield();
}

Text* Parser::parse_section_text(){
	if(it==end){
		return NULL;
	}

	SaneString::iterator start=it;
	char c;

	do{
		c=parse_escape(*it,SECTION_EOF);
	}while(!(c==OPEN_TAG or c==CLOSE_SECTION) and ++it!=end);

	if(start==it){
		return NULL;
	}

	return new Text(text.substr(start-text.begin(),it-start));
}

Section* Parser::parse_section(){
	if(!maybe_char(OPEN_SECTION)){
		return NULL;
	}

	SafePointer<Section> section(new Section());

	//These aren't safe pointers because the moment they have a value,
	//they're immediately handed to another owner.
	Text* text;
	Tag* tag;
	do{
		if(text=parse_section_text()){
			section->add(text);
		}
		if(tag=parse_tag()){
			section->add(tag);
		}
	}while(text or tag);

	if(!maybe_char(CLOSE_SECTION)){
		if(it==end){
			throw ParseError(SECTION_EOF,line,col);
		}
		throw ParseError(EXPECTED_CLOSE_SECTION,line,col);
	}

	return section.yield();
}

Text* Parser::parse_text(){
	if(it==end){
		return NULL;
	}

	SaneString::iterator start=it;
	char c,q;

	if(maybe_char(QUOTE1)){
		q=QUOTE1;
	}
	else if(maybe_char(QUOTE2)){
		q=QUOTE2;
	}
	else if(maybe_char(QUOTE3)){
		q=QUOTE3;
	}
	//Try unquoted
	else{
		do{
			c=parse_escape(*it,UNQUOTED_EOF);
		}while(!(
			//Ordered by expected frequency
			isspace(c) or c==OPEN_TAG or c==CLOSE_TAG or c==NAMED_ATTR or
			c==QUOTE1 or c==QUOTE2 or c==OPEN_SECTION or c==CLOSE_SECTION or
			c==QUOTE3
		) and ++it!=end);

		if(it==start){
			return NULL;
		}

		return new Text(text.substr(start-text.begin(),it-start));
	}

	//Must be quoted
	do{
		c=parse_escape(*it,QUOTED_EOF);
	}while(c!=q and ++it!=end);

	if(!maybe_char(q)){
		throw ParseError(QUOTED_EOF,line,col);
	}

	return new Text(
		((SaneString)text.substr(start-text.begin(),it-start)).escape(q)
	);
}

Flake* Parser::parse_value(){
	Flake* f;
	if(f=parse_text()){
		return f;
	}

	if(f=parse_tag()){
		return f;
	}

	if(f=parse_section()){
		return f;
	}

	//Error checking

	/*
	Snow errors are very predictable, so check for common
	mistakes. By this point, we know the next character is
	one of the quote characters, ], }, whitespace, a control
	character, or EOF (if not, something is HORRIBLY wrong)
	*/

	if(it==end){
		throw ParseError(TAG_EOF,line,col);
	}

	char c=*it;

	if(c==CLOSE_SECTION){
		throw ParseError(UNEXPECTED_CLOSE_SECTION,line,col-1);
	}

	if(c==CLOSE_TAG){
		throw ParseError(UNNAMED_ATTR,colonline,coloncol);
	}

	if(c==NAMED_ATTR){
		throw ParseError(ILLEGAL_NAMED,line,col-1);
	}

	if(isspace(c)){
		//This should NEVER happen. Guarantees a problem with the parser.
		throw ParseError(UNEXPECTED_SPACE,line,col);
	}

	//Reserved for cosmic ray errors
	throw ParseError(COSMIC_RAY_ERR,line,col);
}

Text* Parser::parse_doc_text(){
	if(it==end){
		return NULL;
	}

	SaneString::iterator start=it;
	char c;

	do{
		c=parse_escape(*it,NULL);
	}while(c!=OPEN_TAG and ++it!=end);

	if(start==it){
		return NULL;
	}

	return new Text(text.substr(start-text.begin(),it-start));
}

Parser::Parser(Tagset* ts):ts(ts){}

Document* Parser::parse(const SaneString& s){
	text=s;
	it=text.begin();
	end=text.end();
	line=1;
	col=0;

	SafePointer<Document> doc(new Document());

	//These aren't safe pointers because the moment they have a value,
	//they're immediately handed to another owner.
	Flake* text;
	Flake* tag;
	do{
		if(text=parse_doc_text()){
			doc->add(text);
		}
		if(tag=parse_tag()){
			doc->add(tag);
		}
	}while(text or tag);

	return doc.yield();
}

File: java/Parser.java
package org.snowlang.snow;

/**
 * Public interface for the Snow parser.
**/
public Parser{
	//Error messages
	private const static String SECTION_EOF=
		"Reached end of string/file while parsing a section.";
	private const static String UNQUOTED_EOF=
		"Reached end of string/file while parsing unquoted text.";
	private const static String QUOTED_EOF=
		"Reached end of string/file while parsing quoted text.";
	private const static String EXPECTED_CLOSE_SECTION=
		"Expected close bracket ].";
	private const static String TAG_EOF=
		"Reached end of string/file while parsing a tag";
	private const static String UNEXPECTED_CLOSE_SECTION=
		"Unexpected close bracket ]. Did you forget to close a tag?";
	private const static String UNNAMED_ATTR=
		"Forgot to assign a value to the named attribute.";
	private const static String ILLEGAL_NAMED=
		"The colon is disallowed in unquoted text.";
	private const static String UNEXPECTED_SPACE=
		"Expected a value, found whitespace. "+
		"There's a problem with the API's parser code.";
	private const static String COSMIC_RAY_ERR=
		"Something went horribly wrong. "+
		"There's a problem with the API's parser code.";

	private const static String UNICODE_EOF=
		"Reached end of string/file while parsing a Unicode codepoint.";

	//Relevant characters
	private const static char OPEN_TAG='{';
	private const static char NAMED_ATTR=':';
	private const static char CLOSE_TAG='}';
	private const static char OPEN_SECTION='[';
	private const static char CLOSE_SECTION=']';
	private const static char QUOTE1='"';
	private const static char QUOTE2='\'';
	private const static char QUOTE3='`'
	private const static char ESCAPE_CHAR='\\';
	
	//Character escape strings
	private const static String UNQUOTED_ESC=
		ESCAPE_CHAR+" \n\r\t\v\f\u0085\u00a0\u1680"+
		OPEN_TAG+NAMED_ATTR+CLOSE_TAG+OPEN_SECTION+CLOSE_SECTION+
		"\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009"+
		"\u200a\u2028\u2029\u202f\u205f\u3000";
	private const static String QUOTED_ESC=""+ESCAPE_CHAR;
	private const static String DOCUMENT_ESC=""+ESCAPE_CHAR+OPEN_TAG;
	private const static String SECTION_ESC=
		""+ESCAPE_CHAR+OPEN_TAG+CLOSE_SECTION;
	
	/**
	 * Used for giving the location of parsing errors.
	**/
	protected int line,col;
	/**
	 * Positions used for special-case error messages involving colons.
	**/
	protected int colonline,coloncol;
	/**
	 * A general-purpose character iterator used in parsing.
	**/
	protected Utf8Iterator it;
	
	/**
	 * The tagset to use with the parser.
	**/
	protected Tagset ts;
	
	public Parser(Tagset ts){
		this.ts=ts;
	}
	
	/**
	 * Calculate the line and column incrementation based on the given
	 *  character as well as abstract non-\n newline modes. This accepts
	 *  all of the Unicode-specified line terminators.
	 *
	 * @param c The character with which to start processing.
	 *
	 * @return The reinterpreted character (either c or \n).
	**/
	protected char calc_line(char c){
		if(c=='\r'){
			//Windows-style newline \r\n
			if(it.hasNext() && it.peek()=='\n'){
				it.incr();
			}
			//Mac-style newline \r
			++line;
			col=0;
		}
		else if(
			//Unix-style newline
			c=='\n' ||
			//Defined by Unicode as newlines for reasons
			c=='\v' || c=='\f' ||
			//Unicode chars NEL (NExt Line), LS (Line Separator), and
			// PS (Paragraph Separator)
			c=='\u0085' || c=='\u2028' || c=='\u2029'
		){
			++line;
			col=0;
		}
		else{
			++col;
			return c;
		}
		
		return '\n';
	}
	
	/**
	 * Main text parsing function which handles escapes.
	 *
	 * @param text The string builder to use.
	 * @param eof The error message to give if the iterator's end is
	 *  encountered, null if no error should be thrown.
	 * @param esc A string containing characters to escape.
	 *
	 * @return The next character (for determining if the parse loop
	 *  should terminate).
	**/
	protected char parse_escape(StringBuilder sb,String eof,String esc){
		char c=it.next();
		if(c==ESCAPE_CHAR){
			//Escapes consume the next character regardless of what it is
			if(it.hasNext()){
				c=it.next();
				if(esc.indexOf(c)<0){
					sb.append(ESCAPE_CHAR);
				}
				//...but we still need to check for newlines for error
				// messages.
				sb.append(calc_line(c));
				if(it.hasNext()){
					c=it.next();
					sb.append(c);
					return c;
				}
			}
			
			if(eof){
				throw new ParseError(eof,line,col);
			}
			
			//Return doesn't strictly matter, but the OPEN_TAG
			// character is checked before EOF, so this avoids a check.
			return OPEN_TAG;
		}
		
		c=calc_line(c);
		sb.append(c);
		
		return c;
	}
	
	/**
	 * Try to parse a single character and increment on success.
	**/
	protected boolean maybe_char(char c){
		if(it.peek()==c){
			it.incr();
			return true;
		}
		
		return false;
	}
	
	/**
	 * Parse a Snow text object (quoted or unquoted)
	 *
	 * @return The Snow text object or null if empty.
	**/
	protected Text parse_text(){
		if(!it.hasNext()){
			return null;
		}

		StringBuilder text=new StringBuilder();
		char c,q;

		if(maybe_char(QUOTE1)){
			q=QUOTE1;
		}
		else if(maybe_char(QUOTE2)){
			q=QUOTE2;
		}
		else if(maybe_char(QUOTE3)){
			q=QUOTE3;
		}
		//Try unquoted
		else{
			do{
				c=parse_escape(text,UNQUOTED_EOF,UNQUOTED_ESC);
			}while(!(
				//Ordered by expected frequency
				Character.isSpaceChar(c) || c==OPEN_TAG ||
				c==CLOSE_TAG || c==NAMED_ATTR || c==QUOTE1 ||
				c==QUOTE2 || c==OPEN_SECTION || c==CLOSE_SECTION ||
				c==QUOTE3
			) && it.hasNext());

			if(text.length()==0){
				return null;
			}

			return new Text(text.toString());
		}

		//Must be quoted
		do{
			c=parse_escape(text,QUOTED_EOF,QUOTED_ESC+q);
		}while(c!=q && it.hasNext());

		if(!maybe_char(q)){
			throw new ParseError(QUOTED_EOF,line,col);
		}

		return new Text(text.toString());
	}
	
	/**
	 * Parse the text pattern used by sections.
	**/
	protected Text parse_section_text(){
		if(!it.hasNext()){
			return null;
		}

		StringBuilder text=new StringBuilder();
		char c;

		do{
			c=parse_escape(text,SECTION_EOF,SECTION_ESC);
		}while(!(c==OPEN_TAG || c==CLOSE_SECTION) && it.hasNext());

		if(text.length()==0){
			return null;
		}

		return new Text(text.toString());
	}
	
	/**
	 * Parse a Snow section object.
	**/
	protected Section parse_section(){
		if(!maybe_char(OPEN_SECTION)){
			return null;
		}

		Section section=new Section();
		
		Text text;
		Tag tag;
		
		do{
			if(text=parse_section_text()){
				section.add(text);
			}
			if(tag=parse_tag()){
				section.add(tag);
			}
		}while(text || tag);

		if(!maybe_char(CLOSE_SECTION)){
			if(!it.hasNext()){
				throw new ParseError(SECTION_EOF,line,col);
			}
			throw new ParseError(EXPECTED_CLOSE_SECTION,line,col);
		}

		return section;
	}
	
	/**
	 * Parse a Snow value (tag, section, or text).
	 *
	 * This is where most parse errors are processed.
	**/
	protected Flake parse_value(){
		Flake f;
		if(f=parse_text()){
			return f;
		}
		
		if(f=parse_tag()){
			return f;
		}
		
		if(f=parse_section()){
			return f;
		}
		
		//Error checking

		/*
		Snow errors are very predictable, so check for common
		mistakes. By this point, we know the next character is
		one of the quote characters, ], }, whitespace, a control
		character, or EOF (if not, something is HORRIBLY wrong)
		*/
		
		if(!it.hasNext()){
			throw new ParseError(TAG_EOF,line,col);
		}
		
		char c=it.next();
		
		if(c==CLOSE_SECTION){
			throw new ParseError(UNEXPECTED_CLOSE_SECTION,line,col-1);
		}

		if(c==CLOSE_TAG){
			throw new ParseError(UNNAMED_ATTR,colonline,coloncol);
		}

		if(c==NAMED_ATTR){
			throw new ParseError(ILLEGAL_NAMED,line,col-1);
		}

		if(Character.isSpaceChar(c)){
			//This should NEVER happen. Guarantees a problem with the
			// parser.
			throw new ParseError(UNEXPECTED_SPACE,line,col);
		}

		//Reserved for cosmic ray errors
		throw new ParseError(COSMIC_RAY_ERR,line,col);
	}
	
	/**
	 * Parse as much whitespace as possible.
	**/
	protected void space(){
		char c;
		while(it.hasNext() && Character.isSpaceChar(c=it.next())){
			calc_line(c);
		}
	}
	
	/**
	 * Try to parse a tag in the Snow document, else return null.
	**/
	protected Tag parse_tag(){
		if(!maybe_char(OPEN_TAG)){
			return null;
		}
		space();
		
		Tag tag=new Tag();
		
		while(!maybe_char(CLOSE_TAG)){
			Flake key=parse_value();
			space();
			if(maybe_char(NAMED_ATTR)){
				colonline=line;
				coloncol=col-1;
				
				space();
				Flake val=parse_value();
				space();
				
				Flake f=tag.get(key);
				if(f){
					Section s=f.as_section();
					if(s){
						s.add(val);
					}
					else{
						tag.set(key,new Section(f,val));
					}
				}
				else{
					tag.set(key,val);
				}
			}
			else{
				tag.add(key);
			}
		}
		
		if(ts){
			Tagdef td=ts.get(tag.get(0));
			if(td){
				return td.process(tag).as_tag();
			}
		}
		return tag;
	}
	
	/**
	 * Parse the character pattern for a document's top level.
	**/
	protected Text parse_doc_text(){
		if(!it.hasNext()){
			return null;
		}

		StringBuilder text=new StringBuilder();
		char c;

		do{
			c=process_escape(text,null,DOCUMENT_ESC);
		}while(c!=OPEN_TAG && it.hasNext());

		if(text.length()==0){
			return null;
		}

		return new Text(text.toString());
	}
	
	/**
	 * Parse a Snow document using the given iterator.
	**/
	public Document parse(Utf8Iterator cit)  ParseError{
		line=1;
		col=0;
		it=cit;
		end=cit.end();
		
		Document doc=new Document();
		Text text=null,tag=null;
		
		do{
			if(text=parse_doc_text()){
				doc.add(text);
			}
			if(tag=parse_tag()){
				doc.add(tag);
			}
		}while(text || tag);
		
		return doc;
	}
	
	/**
	 * Parse a Snow document using the given string.
	**/
	public Document parse(String s){
		return parse(new StringUtf8Iterator(s));
	}
}